llvm/lib/IR/RepoTicket.cpp

```diff
@@ -118,76 +127,122 @@ static GODigestState accumulateGOFinalDigestOrInitialDigestAndGetGODigestState(
+  // Update the GO's hash using its contributions. Contributions don't form the
+  // loop since 1) only the global variable has the contributions and 2) the
+  // contributions are the functions.
+  for (const GlobalObject *const G : State.Contributions) {
+    Changed = true;
+    GOHash.update(GOIMap[G].InitialDigest.Bytes);
+  }
+  // Update the GO's hash using its dependencies.
+  auto LoopPoint = std::numeric_limits<size_t>::max();
+  for (const GlobalObject *const G : State.Dependencies) {
     const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
     // if function will not be inlined and not be discarded if it is not used,
     // skip it.
     if (Fn && Fn->hasFnAttribute(Attribute::NoInline) &&
         !Fn->isDiscardableIfUnused())
       continue;
-    Changed = updateDigestUseDependenciesAndContributions(
-                  G, GOHash, Visited, GOIMap, AccumulateGODigest) ||
-              Changed;
+    const auto AdjGO = updateDigestUseDependenciesAndContributions(
+        G, Visited, GOIMap, HashTable, AccumulateGODigest);
+    Changed = Changed || std::get<0>(AdjGO);
+    // A GO which is loop back to itself doesn't count as a loop.
+    if (G != GO) {
+      LoopPoint = std::min(LoopPoint, std::get<1>(AdjGO));
+    }
+    GOHash.update(std::get<2>(AdjGO).Bytes);
   }
+  // Encoded the final edge. Record that in the hash.
+  GOHash.update(static_cast<uint8_t>(Tags::End));
 
-  return Changed;
+  MD5::MD5Result Digest;
+  GOHash.final(Digest);
+  auto const result = std::make_tuple(Changed, LoopPoint, Digest);
+
+  if (LoopPoint > NumVisited) {
+    HashTable[GO] = Digest;
+    LLVM_DEBUG(dbgs() << "Recording result for \"" << GO->getName() << "\"\n");
+    ++NumMemoizedHashes;
+    ++VisitedCachedGOTimes;
+  } else {
+    ++VisitedUnCachedGOTimes;
+  }
+  return result;
 }
 
-// Calculate the GOs' initial digest, dependencies, contributions and GONumber.
-std::pair<GOInfoMap, GONumber> calculateGONumAndGOIMap(Module &M) {
+// Calculate the GOs' initial digest, dependencies, contributions and the number
+// of hashed global variables and functions.
+GOInfoMap calculateGONumAndGOIMap(Module &M) {
   GOInfoMap GOIMap;
-  GONumber GONum;
 
   for (auto &GV : M.globals()) {
     if (GV.isDeclaration())
       continue;
     calculateGOInfo(&GV, GOIMap);
-    ++GONum.VarNum;
+    ++NumVariables;
   }
   for (auto &Fn : M.functions()) {
     if (Fn.isDeclaration())
       continue;
     calculateGOInfo(&Fn, GOIMap);
-    ++GONum.FuncNum;
+    ++NumFunctions;
   }
-  return std::make_pair(std::move(GOIMap), std::move(GONum));
+  return std::move(GOIMap);
 }
 
-std::tuple<bool, unsigned, unsigned> generateTicketMDs(Module &M) {
+bool generateTicketMDs(Module &M) {
   bool Changed = false;
-  GONumber GONum;
   GOStateMap Visited;
   Visited.reserve(M.size() + M.getGlobalList().size());
   GOInfoMap GOIMap;
+  MemoizedHashes HashTable;
+  HashTable.reserve(M.size() + M.getGlobalList().size());
```

Comment:
This repeats the calculation "M.size() + M.getGlobalList().size()". Could this be factored out and named?

Reply:
Factored, thanks.